helix-parsec/src/lib.rs

```diff
@@ -0,0 +1,580 @@
+            Err(_) => Err($input),
+        }
+    };
+    ($input:expr, $next_input:expr, $last:ident ; $(,)? $($acc:ident),*) => {
+        match $last.parse($next_input) {
+            Ok((next_input, last)) => Ok((next_input, ($($acc),+, last))),
+            Err(_) => Err($input),
+        }
+    }
+}
+
+seq_impl!(A, B);
+seq_impl!(A, B, C);
+seq_impl!(A, B, C, D);
+seq_impl!(A, B, C, D, E);
+seq_impl!(A, B, C, D, E, F);
+seq_impl!(A, B, C, D, E, F, G);
+seq_impl!(A, B, C, D, E, F, G, H);
+seq_impl!(A, B, C, D, E, F, G, H, I);
+seq_impl!(A, B, C, D, E, F, G, H, I, J);
+
+/// A parser combinator which chooses the first of the input parsers which matches
+/// successfully.
+///
+/// All input parsers must have the same output type. This is a variadic form for [or].
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{choice, or, Parser};
+/// let parser = choice!("foo", "bar", "baz");
+/// assert_eq!(Ok(("", "foo")), parser.parse("foo"));
+/// assert_eq!(Ok(("", "bar")), parser.parse("bar"));
+/// assert_eq!(Err("quiz"), parser.parse("quiz"));
+/// ```
+#[macro_export]
+macro_rules! choice {
+    ($parser: expr $(,)?) => {
+        $parser
+    };
+    ($parser: expr, $($rest: expr),+ $(,)?) => {
+        or($parser, choice!($($rest),+))
+    }
+}
+
+// Ordinary parser combinators
+
+/// A parser combinator which takes a parser as input and maps the output using the
+/// given transformation function.
+///
+/// This corresponds to [Result::map]. The value is only mapped if the input parser
+/// matches against input.
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{map, Parser};
+/// let parser = map("123", |s| s.parse::<i32>().unwrap());
+/// assert_eq!(Ok(("", 123)), parser.parse("123"));
+/// assert_eq!(Err("abc"), parser.parse("abc"));
+/// ```
+pub fn map<'a, P, F, T>(parser: P, map_fn: F) -> impl Parser<'a, Output = T>
+where
+    P: Parser<'a>,
+    F: Fn(P::Output) -> T,
+{
+    move |input| {
+        parser
+            .parse(input)
+            .map(|(next_input, result)| (next_input, map_fn(result)))
+    }
+}
+
+/// A parser combinator which succeeds if the given parser matches the input and
+/// the given `filter_map_fun` returns `Some`.
+///
+/// # Examples
+///
+/// ```
+/// use helix_parsec::{filter_map, take_until, Parser};
+/// let parser = filter_map(take_until(|c| c == '.'), |s| s.parse::<i32>().ok());
+/// assert_eq!(Ok((".456", 123)), parser.parse("123.456"));
+/// assert_eq!(Err("abc.def"), parser.parse("abc.def"));
+/// ```
+pub fn filter_map<'a, P, F, T>(parser: P, filter_map_fn: F) -> impl Parser<'a, Output = T>
+where
+    P: Parser<'a>,
+    F: Fn(P::Output) -> Option<T>,
+{
+    move |input| match parser.parse(input) {
+        Ok((next_input, value)) => match filter_map_fn(value) {
+            Some(value) => Ok((next_input, value)),
+            None => Err(input),
+        },
+        Err(_) => Err(input),
+    }
+}
+
+/// TODO: name
```

Comment:
```suggestion
```

This was mostly a note-to-self. I'm ok with reparse_as now that I think about it more